src/cdi_ipcress/python/ipcress_reader.py

```diff
@@ -1,327 +1,337 @@
+    n_rho = len(rho_grid)
+    # n_T = len(T_grid)
+    n_hnu = len(hnu_grid)
+
+    # don't allow extrapolation
+    if (target_rho < np.min(rho_grid)):
+        target_rho = np.min(rho_grid)
+    if (target_rho > np.max(rho_grid)):
+        target_rho = np.max(rho_grid)
+    if (target_T < np.min(T_grid)):
+        target_T = np.min(T_grid)
+    if (target_T > np.max(T_grid)):
+        target_T = np.max(T_grid)
+    if (print_str is not None):
+        print("Interpolating {0}--Target rho: {1} , target T: {2}".format(
+            print_str, target_rho, target_T))
+
+    # get correct index of adjacent density points
+    rho_L = 1000
+    rho_G = 0
+    for rho_i, rho in enumerate(rho_grid[:-1]):
+        if (target_rho >= rho and target_rho <= rho_grid[rho_i + 1]):
+            rho_L = rho_i
+            rho_G = rho_i + 1
+            break
+
+    # get correct index of adjacent temperature points
+    T_L = 1000
+    T_G = 0
+    for T_i, T in enumerate(T_grid[:-1]):
+        if (target_T >= T and target_T <= T_grid[T_i + 1]):
+            T_L = T_i
+            T_G = T_i + 1
+            break
+
+    # print("Temperature interpolation bounds: {0} {1}".format(T_grid[T_L], T_grid[T_G]))
+    # print("Density interpolation bounds: {0} {1}".format(rho_grid[rho_L], rho_grid[rho_G]))
+
+    # get the adjacent rows of the opacity index
+    # get the points of the opacity index
+    rho_L_T_L = op_data[n_rho * T_L * (n_hnu - 1) + rho_L * (n_hnu - 1):n_rho * T_L * (n_hnu - 1) +
+                        rho_L * (n_hnu - 1) + (n_hnu - 1)]
+    rho_L_T_G = op_data[n_rho * T_G * (n_hnu - 1) + rho_L * (n_hnu - 1):n_rho * T_G * (n_hnu - 1) +
+                        rho_L * (n_hnu - 1) + (n_hnu - 1)]
+    rho_G_T_L = op_data[n_rho * T_L * (n_hnu - 1) + rho_G * (n_hnu - 1):n_rho * T_L * (n_hnu - 1) +
+                        rho_G * (n_hnu - 1) + (n_hnu - 1)]
+    rho_G_T_G = op_data[n_rho * T_G * (n_hnu - 1) + rho_G * (n_hnu - 1):n_rho * T_G * (n_hnu - 1) +
+                        rho_G * (n_hnu - 1) + (n_hnu - 1)]
+
+    interp_op = []
+    # interpolate for each frequency point
+    for i in range(n_hnu - 1):
+        # if (rho_L_T_L[i] < 1.0e-10) or (rho_L_T_G[i] < 1.0e-10) or (rho_G_T_L[i] < 1.0e-10) or
+        # (rho_G_T_G[i] < 1.0e10):
+        #    interp_op.append(1.0e-10)
+        # print("{0} {1} {2} {3}" .format(rho_L_T_L[i], rho_L_T_G[i], rho_G_T_L[i], rho_G_T_G[i]))
+        log_op_T_L = log(rho_L_T_L[i]) + log(target_rho / rho_grid[rho_L]) / \
+            log(rho_grid[rho_G] / rho_grid[rho_L]) * log(rho_G_T_L[i] / rho_L_T_L[i])
+        log_op_T_G = log(rho_L_T_G[i]) + log(target_rho / rho_grid[rho_L]) / \
+            log(rho_grid[rho_G] / rho_grid[rho_L]) * log(rho_G_T_G[i] / rho_L_T_G[i])
+        log_op = log_op_T_L + log(target_T / T_grid[T_L]) / log(T_grid[T_G] / T_grid[T_L]) * \
+            (log_op_T_G - log_op_T_L)
+        interp_op.append(exp(log_op))
+
+    print("hnu(keV)            opacity(sq_cm/g)         opacity(1/cm)")
```

Comment:
Maybe better for a future commit, but could we put a line 68-style `if (print_str is not None):` so lines 119-122 aren't verbosely printed in all cases?

Reply:
I can do this in a follow-on PR since I want to clean up some of the unused import statements.

Reply:
Sounds good. You might want to avoid things like `from numpy import *` too